crates/futures/src/lib.rs

```diff
@@ -7,46 +7,222 @@
+                drop(state.callbacks.take());
+
+                // Next, store the value into the internal state.
+                state.result = Some(val);
+                state.task.take()
+            };
+
+            // And then finally if any task was waiting on the value wake it up and
+            // let them know it's there.
+            if let Some(task) = task {
+                task.wake()
+            }
+        }
+
+        let resolve = {
+            let state = state.clone();
+            Closure::once(move |val| finish(&state, Ok(val)))
+        };
+
+        let reject = {
+            let state = state.clone();
+            Closure::once(move |val| finish(&state, Err(val)))
+        };
+
+        js.then2(&resolve, &reject);
+
+        state.borrow_mut().callbacks = Some((resolve, reject));
+
+        JsFuture { inner: state }
+    }
+}
+
+impl Future for JsFuture {
+    type Output = Result<JsValue, JsValue>;
+
+    fn poll(self: Pin<&mut Self>, cx: &mut Context) -> Poll<Self::Output> {
+        let mut inner = self.inner.borrow_mut();
+
+        // If our value has come in then we return it...
+        if let Some(val) = inner.result.take() {
+            return Poll::Ready(val);
+        }
+
+        // ... otherwise we arrange ourselves to get woken up once the value
+        // does come in
+        inner.task = Some(cx.waker().clone());
+        Poll::Pending
+    }
+}
+
+/// Converts a Rust `Future` into a JavaScript `Promise`.
+///
+/// This function will take any future in Rust and schedule it to be executed,
+/// returning a JavaScript `Promise` which can then be passed to JavaScript.
+///
+/// The `future` must be `'static` because it will be scheduled to run in the
+/// background and cannot contain any stack references.
+///
+/// The returned `Promise` will be resolved or rejected when the future completes,
+/// depending on whether it finishes with `Ok` or `Err`.
+///
+/// # Panics
+///
+/// Note that in wasm panics are currently translated to aborts, but "abort" in
+/// this case means that a JavaScript exception is thrown. The wasm module is
+/// still usable (likely erroneously) after Rust panics.
+///
+/// If the `future` provided panics then the returned `Promise` **will not
+/// resolve**. Instead it will be a leaked promise. This is an unfortunate
+/// limitation of wasm currently that's hoped to be fixed one day!
+pub fn future_to_promise<F>(future: F) -> Promise
+where
+    F: Future<Output = Result<JsValue, JsValue>> + 'static,
+{
+    let mut future = Some(future);
 
-mod shared;
-pub use shared::*;
+    Promise::new(&mut |resolve, reject| {
+        let future = future.take().unwrap_throw();
 
-cfg_if! {
-    if #[cfg(target_feature = "atomics")] {
-        mod wait_async_polyfill;
-        mod multithread;
-        pub use multithread::*;
-    } else {
-        mod singlethread;
-        pub use singlethread::*;
-     }
+        // TODO change Promise::new to be FnOnce
```

Comment:
I don't think `#[wasm_bindgen]` supports this, so I'd prefer the `TODO` be removed an an issue filed (although I suspect there's already an issue)